src/linear_algebra/lagrange.lean

```diff
@@ -1,224 +1,371 @@
+  exact vandermonde_agreement (function.embedding.trans (fintype.equiv_fin ι).symm.to_embedding v)
+                              hfg1 (λ _, hfg2 _)
+end
+
+theorem eq_of_eval_eq' (v : ι ↪ F) {f g : F[X]}
+  (hf : f.degree < fintype.card ι) (hg : g.degree < fintype.card ι)
+  (hfg : ∀ i, f.eval (v i) = g.eval (v i)) : f = g :=
+begin
+  rw ← mem_degree_lt at hf hg,
+  exact vandermonde_agreement'  (function.embedding.trans (fintype.equiv_fin ι).symm.to_embedding v)
+                                hf hg (λ _, hfg _)
+end
+
+end non_lagrange
 
 /-!
 # Lagrange interpolation
 
 ## Main definitions
-
-* `lagrange.basis s x` where `s : finset F` and `x : F`: the Lagrange basis polynomial
-  that evaluates to `1` at `x` and `0` at other elements of `s`.
-* `lagrange.interpolate s f` where `s : finset F` and `f : F → F`: the Lagrange interpolant
-  that evaluates to `f x` at `x` for `x ∈ s`.
+* In everything that follows, `v : ι ↪ F` is an embedding to the field from a fintype `ι`.
+Conceptually, this is a set of distinct nodes around which we interpolate.
+* `lagrange.basis_divisor v i j`, with `i j : ι`. These are the normalised irreducible factors of
+the Lagrange basis polynomials. They evaluate to `1` at `v i` and `0` at `v j` when `i` and `j`
+are distinct.
+* `lagrange.basis v i` with `i : ι`: the Lagrange basis polynomial that evaluates to `1` at `v i`  and `0` at `v j` for `i ≠ j`.
+* `lagrange.interpolate v r` where `r : ι → F` is a function from the fintype to the field: the Lagrange interpolant that evaluates to `r i` at `x i` for all `i : ι`. The `r i` are the _values_
+associated with the _nodes_`x i`.
+* `lagrange.interpolate_at v f`, where `v : ι ↪ F` and `ι` is a fintype, and `f : F → F` is a
+function from the field to itself: this is the Lagrange interpolant that evaluates to `f (x i)`
+at `x i`, and so approximates the function `f`. This is just a special case of the general
+interpolation, where the values are given by a known function `f`.
 -/
 
 noncomputable theory
-open_locale big_operators classical polynomial
-
-universe u
+open_locale big_operators polynomial
 
 namespace lagrange
-
-variables {F : Type u} [decidable_eq F] [field F] (s : finset F)
-variables {F' : Type u} [field F'] (s' : finset F')
+open finset
+variables {F : Type*} [field F]
 
 open polynomial
 
-/-- Lagrange basis polynomials that evaluate to 1 at `x` and 0 at other elements of `s`. -/
-def basis (x : F) : F[X] :=
-∏ y in s.erase x, C (x - y)⁻¹ * (X - C y)
+section node
+variables {ι : Type*} {v : ι ↪ F} {i j : ι}
+/-- The basis divisor is defined in terms of an embedded `v : ι ↪ F` from a fintype `ι`. `basis_divisor v i j` is the unique polynomial with `degree ≤ 1` such that when evaluated at `v i` it gives `1` and `v j` it gives `0` (where when `i = j` it is identically `0`).
+
+Conceptually, they are therefore the building blocks for the Lagrange interpolants. -/
+def basis_divisor (v : ι ↪ F) : ι → ι → F[X] := λ i j, C (v i - v j)⁻¹ * (X - C (v j))
```

Comment:
```suggestion
Conceptually, they are therefore the building blocks for the Lagrange interpolants. -/
def basis_divisor (v : ι ↪ F) (i j : ι) : F[X] := C (v i - v j)⁻¹ * (X - C (v j))
```
then you can remove `basis_divisor_def` below as it's just `basis_divisor` (the generated lemma)